### PR TITLE
JAMES-2806 Add BucketName param into BlobStore API

### PR DESCRIPTION
**Pull request by trantienduchn**
This is the base for Memory & ObjectStorage implementation. It's suppose to be simple enough to have a base to work in parallel later

### CODE REVIEW

**Comment by trantienduchn** [comment]
```
[d67af8198a1dd8f20f1ff1c559ecb7ecf0855132] [ERROR] Errors: 
[d67af8198a1dd8f20f1ff1c559ecb7ecf0855132] [ERROR]   AttachmentV2MigrationTest.runShouldReturnPartialWhenSavingBlobsFails:158 InvalidUseOfMatchers
```

**Comment by trantienduchn** [comment]
```
[078c1942fd6526d45974a0715547fbe1b26abb72] [ERROR]   DiscreteDistributionTest.partitionShouldSupportDuplicatedDistributionEntry:112 
[078c1942fd6526d45974a0715547fbe1b26abb72] Expecting:
[078c1942fd6526d45974a0715547fbe1b26abb72]   <668377L>
[078c1942fd6526d45974a0715547fbe1b26abb72] to be close to:
[078c1942fd6526d45974a0715547fbe1b26abb72]   <663246L>
[078c1942fd6526d45974a0715547fbe1b26abb72] by less than <5000L> but difference was <5131L>.
[078c1942fd6526d45974a0715547fbe1b26abb72] (a difference of exactly <5000L> being considered valid)
```
test this please

File: server/blob/blob-api/src/main/java/org/apache/james/blob/api/BucketName.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    private BucketName(String value) {
+        this.bucket = value;
+    }
+
+    public String value() {
+        return bucket;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (chibenwa) [inline server/blob/blob-api/src/main/java/org/apache/james/blob/api/BucketName.java]: Please use the standard format for this:

```
    @Override
    public final boolean equals(Object o) {
        if (o instanceof BucketName) {
            BucketName that = (BucketName) o;

            return Objects.equals(this.value, that.value);
        }
        return false;
    }
```

And let's get rid of micro - optimization...

File: server/blob/blob-api/src/main/java/org/apache/james/blob/api/BucketName.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    private final String bucket;
+
+    private BucketName(String value) {
+        this.bucket = value;
+    }
+
+    public String value() {
+        return bucket;
```
Comment (mbaechler) [inline server/blob/blob-api/src/main/java/org/apache/james/blob/api/BucketName.java]: asString or serialize would be better?

File: server/blob/blob-api/src/test/java/org/apache/james/blob/api/BucketNameTest.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.blob.api;
+
+import org.junit.jupiter.api.Test;
+
+import nl.jqno.equalsverifier.EqualsVerifier;
+
+class BucketNameTest {
```
Comment (mbaechler) [inline server/blob/blob-api/src/test/java/org/apache/james/blob/api/BucketNameTest.java]: I'm sure you want to constrain the names a bit, like for example, not empty

**Comment by chibenwa** [comment]
I would like you to:
 - reject "non-default Buckets" in all 3 implementations
 - Create the default bucket in ObjectStorageDependenciesModule::buildObjectStore
 - Get rid of `ObjectStorageBlobConfiguration::getNamespace` as we will always use the bucket named default (as far as I understood)

**Comment by trantienduchn** [comment]
> reject "non-default Buckets" in all 3 implementations

Will be taken into account in next PRs
> Create the default bucket in ObjectStorageDependenciesModule::buildObjectStore
  Get rid of ObjectStorageBlobConfiguration::getNamespace as we will always use the bucket named default (as far as I understood)

I will do it in the PR of updating ObjectStorageBlobStore

File: server/blob/blob-api/src/test/java/org/apache/james/blob/api/BucketNameTest.java (hunk 3 above)
Comment (trantienduchn) [reply server/blob/blob-api/src/test/java/org/apache/james/blob/api/BucketNameTest.java]: I did it, in fixup commits

File: server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAO.java [hunk 4]
```diff
@@ -92,12 +93,12 @@
...
     @Override
-    public Mono<BlobId> save(byte[] data) {
-        return save(new ByteArrayInputStream(data));
+    public Mono<BlobId> save(BucketName bucketName, byte[] data) {
+        return save(bucketName, new ByteArrayInputStream(data));
     }
 
     @Override
-    public Mono<BlobId> save(InputStream data) {
+    public Mono<BlobId> save(BucketName bucketName, InputStream data) {
```
Comment (remk) [inline server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAO.java]: We don't use the bucket name ?

Comment (remk) [reply server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAO.java]: ah ok it's a preliminary work :)

Comment (Arsnael) [reply server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAO.java]: Yes it is :)

**Comment by trantienduchn** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**